Every year, American high school students take SATs, which are standardized tests intended to measure literacy, numeracy, and writing skills. There are three sections - reading, math, and writing, each with a **maximum score of 800 points**. These tests are extremely important for students and colleges, as they play a pivotal role in the admissions process.

Analyzing the performance of schools is important for a variety of stakeholders, including policy and education professionals, researchers, government, and even parents considering which school their children should attend. 

You have been provided with a dataset called `schools.csv`, which is previewed below.

You have been tasked with answering three key questions about New York City (NYC) public school SAT performance.

In [2]:
import pandas as pd

schools = pd.read_csv("schools.csv")

schools.head()

school_name    borough building_code  \
0  New Explorations into Science, Technology and ...  Manhattan          M022   
1                               Essex Street Academy  Manhattan          M445   
2                       Lower Manhattan Arts Academy  Manhattan          M445   
3    High School for Dual Language and Asian Studies  Manhattan          M445   
4      Henry Street School for International Studies  Manhattan          M056   

   average_math  average_reading  average_writing  percent_tested  
0           657              601              601             NaN  
1           395              411              387            78.9  
2           418              428              415            65.1  
3           613              453              463            95.9  
4           410              406              381            59.7

Question: Which NYC schools have the best math results?

* The best math results are at least 80% of the *maximum possible score of 800* for math.
* Save your results in a pandas DataFrame called best_math_schools, including "school_name" and "average_math" columns, sorted by "average_math" in descending order.

In [3]:
schools.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 375 entries, 0 to 374
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   school_name      375 non-null    object 
 1   borough          375 non-null    object 
 2   building_code    375 non-null    object 
 3   average_math     375 non-null    int64  
 4   average_reading  375 non-null    int64  
 5   average_writing  375 non-null    int64  
 6   percent_tested   355 non-null    float64
dtypes: float64(1), int64(3), object(3)
memory usage: 20.6+ KB


In [4]:
# Subset by math scores at least 80% of 800
best_math_schools = schools[schools['average_math'] >= 800*(.80)]

best_math_schools.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, 0 to 365
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   school_name      10 non-null     object 
 1   borough          10 non-null     object 
 2   building_code    10 non-null     object 
 3   average_math     10 non-null     int64  
 4   average_reading  10 non-null     int64  
 5   average_writing  10 non-null     int64  
 6   percent_tested   9 non-null      float64
dtypes: float64(1), int64(3), object(3)
memory usage: 640.0+ bytes


In [5]:
best_math_schools['average_math']

0      657
45     641
68     683
88     754
93     711
170    714
174    669
280    682
333    680
365    701
Name: average_math, dtype: int64

In [6]:
# Save as a DataFrame
best_math_schools = best_math_schools[['school_name', 'average_math']].sort_values(by='average_math', ascending=False)

In [7]:
best_math_schools

school_name  average_math
88                              Stuyvesant High School           754
170                       Bronx High School of Science           714
93                 Staten Island Technical High School           711
365  Queens High School for the Sciences at York Co...           701
68   High School for Mathematics, Science, and Engi...           683
280                     Brooklyn Technical High School           682
333                        Townsend Harris High School           680
174  High School of American Studies at Lehman College           669
0    New Explorations into Science, Technology and ...           657
45                       Eleanor Roosevelt High School           641

Question: What are the top 10 performing schools based on the combined SAT scores?

* Save your results as a pandas DataFrame called top_10_schools containing the "school_name" and a new column named "total_SAT", with results ordered by "total_SAT" in descending order ("total_SAT" being the sum of math, reading, and writing scores).

In [8]:
# Create new column to store total scores
schools['total_SAT'] = (
    schools['average_math'] + 
    schools['average_reading'] + 
    schools['average_writing']
)

schools.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 375 entries, 0 to 374
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   school_name      375 non-null    object 
 1   borough          375 non-null    object 
 2   building_code    375 non-null    object 
 3   average_math     375 non-null    int64  
 4   average_reading  375 non-null    int64  
 5   average_writing  375 non-null    int64  
 6   percent_tested   355 non-null    float64
 7   total_SAT        375 non-null    int64  
dtypes: float64(1), int64(4), object(3)
memory usage: 23.6+ KB


In [11]:
schools.head()

school_name    borough building_code  \
0  New Explorations into Science, Technology and ...  Manhattan          M022   
1                               Essex Street Academy  Manhattan          M445   
2                       Lower Manhattan Arts Academy  Manhattan          M445   
3    High School for Dual Language and Asian Studies  Manhattan          M445   
4      Henry Street School for International Studies  Manhattan          M056   

   average_math  average_reading  average_writing  percent_tested  total_SAT  
0           657              601              601             NaN       1859  
1           395              411              387            78.9       1193  
2           418              428              415            65.1       1261  
3           613              453              463            95.9       1529  
4           410              406              381            59.7       1197

In [12]:
# Create a new df with cols 'school_name' and 'total_score' and sort by SAT
sorted_df = schools[['school_name', 'total_SAT']].sort_values(by='total_SAT', ascending=False)

sorted_df.head()

school_name  total_SAT
88                              Stuyvesant High School       2144
170                       Bronx High School of Science       2041
93                 Staten Island Technical High School       2041
174  High School of American Studies at Lehman College       2013
333                        Townsend Harris High School       1981

In [13]:
# Select only top 10 schools
top_10_schools = sorted_df.head(10)

print(top_10_schools)

                                           school_name  total_SAT
88                              Stuyvesant High School       2144
170                       Bronx High School of Science       2041
93                 Staten Island Technical High School       2041
174  High School of American Studies at Lehman College       2013
333                        Townsend Harris High School       1981
365  Queens High School for the Sciences at York Co...       1947
5                       Bard High School Early College       1914
280                     Brooklyn Technical High School       1896
45                       Eleanor Roosevelt High School       1889
68   High School for Mathematics, Science, and Engi...       1889


Question: Which single borough has the largest standard deviation in the combined SAT score?

* Save your results as a pandas DataFrame called largest_std_dev.
* The DataFrame should contain one row, with:
    * "borough" - the name of the NYC borough with the largest standard deviation of "total_SAT".
    * "num_schools" - the number of schools in the borough.
    * "average_SAT" - the mean of "total_SAT".
    * "std_SAT" - the standard deviation of "total_SAT".
* Round all numeric values to two decimal places.

In [21]:
# Group schools by borough
grouped_by_borough = schools.groupby('borough')

grouped_by_borough.head()

school_name        borough  \
0    New Explorations into Science, Technology and ...      Manhattan   
1                                 Essex Street Academy      Manhattan   
2                         Lower Manhattan Arts Academy      Manhattan   
3      High School for Dual Language and Asian Studies      Manhattan   
4        Henry Street School for International Studies      Manhattan   
89   Ralph R. McKee Career and Technical Education ...  Staten Island   
90                          Michael J. Petrides School  Staten Island   
91                                  Curtis High School  Staten Island   
92                           Port Richmond High School  Staten Island   
93                 Staten Island Technical High School  Staten Island   
99                           New Explorers High School          Bronx   
100      Bronx School for Law, Government, and Justice          Bronx   
101        Urban Assembly School for Careers in Sports          Bronx   
102            Bronx Leadership Academy II High School          Bronx   
103            Urban Assembly Bronx Academy of Letters          Bronx   
197        Queens Vocational and Technical High School         Queens   
198               High School of Applied Communication         Queens   
199  Middle College High School at LaGuardia Commun...         Queens   
200         Academy for Careers in Television and Film         Queens   
201  Aviation Career and Technical Education High S...         Queens   
214  Science Skills Center High School for Science,...       Brooklyn   
215                 Brooklyn International High School       Brooklyn   
216                   School for International Studies       Brooklyn   
217                 Brooklyn School for Global Studies       Brooklyn   
218  City Polytechnic High School of Engineering, A...       Brooklyn   

    building_code  average_math  average_reading  average_writing  \
0            M022           657              601              601   
1            M445           395              411              387   
2            M445           418              428              415   
3            M445           613              453              463   
4            M056           410              406              381   
89           R600           420              429              409   
90           R880           483              473              470   
91           R450           453              458              444   
92           R445           432              427              425   
93           R440           711              660              670   
99           X790           390              398              376   
100          X460           403              409              415   
101          X790           402              405              395   
102          X790           384              355              361   
103          X183           398              400              405   
197          Q600           467              436              432   
198          Q735           435              437              441   
199          Q520           377              389              377   
200          Q404           444              458              444   
201          Q610           511              464              456   
214          K805           399              397              386   
215          K805           432              334              333   
216          K293           417              406              394   
217          K293           393              381              402   
218          K580           503              477              427   

     percent_tested  total_SAT  
0               NaN       1859  
1              78.9       1193  
2              65.1       1261  
3              95.9       1529  
4              59.7       1197  
89             38.0       1258  
90             79.2       1426  
91             64.2       1355  
92             44.6       1284  
93             99.7

In [28]:
# Calculate the standard deviation of 'total_score' of each borough
std_by_borough = grouped_by_borough['total_SAT'].std()

std_by_borough.sort_values(ascending=False)

borough
Manhattan        230.294140
Staten Island    222.303596
Queens           195.252899
Brooklyn         154.868427
Bronx            150.393901
Name: total_SAT, dtype: float64

In [30]:
# Find the borough with largest std
largest_std_borough = std_by_borough.idxmax()

largest_std_borough

'Manhattan'

Answer: 'Manhattan' has the largest standard deviation in the combined SAT score.

In [31]:
# Select all columns for borough with largest std
filtered_borough_data = schools[schools['borough'] == largest_std_borough]

# Calculate the number of schools, average SAT, and standard deviation for that borough
num_schools = filtered_borough_data['borough'].count()
average_SAT = filtered_borough_data['total_SAT'].mean()
std_SAT = filtered_borough_data['total_SAT'].std()

# Round the numeric values to two decimal places
num_schools = round(num_schools, 2)
average_SAT = round(average_SAT, 2)
std_SAT = round(std_SAT, 2)

# Create largest_std_dev dataframe
largest_std_dev = pd.DataFrame({
    'borough': [largest_std_borough],
    'num_schools': [num_schools],
    'average_SAT': [average_SAT],
    'std_SAT': [std_SAT]
})

# Display the result
print(largest_std_dev)

     borough  num_schools  average_SAT  std_SAT
0  Manhattan           89      1340.13   230.29


'Manhattan' borough has total of 89 schools, the average SAT scores at 1340.13, and the standard deviation of SAT score at 230.29. 